src/app/scanning.service.ts

```diff
@@ -0,0 +1,182 @@
+import { Observable, Subscriber, Subscription } from 'rxjs';
+
+export class Scannable {
+  private _priority: number;
+  private _type: string;
+
+  constructor(priority: number, type: string) {
+    this._priority = priority;
+    this._type = type;
+  }
+
+  get priority(): number {
+    return this._priority;
+  }
+
+  set priority(priority: number) {
+    this._priority = priority;
+  }
+
+  get type(): string {
+    return this._type;
+  }
+
+  set type(type: string) {
+    this._type = type;
+  }
+
+  hasChildren(): boolean {
+    return false;
+  }
+}
+
+export class ScannableCollection extends Scannable {
+  private children: Scannable[] = [];
+
+  hasChildren(): boolean {
+    return this.children.length > 0;
+  }
+
+  getChildren(): Scannable[] {
+    return this.children;
+  }
+
+  getChild(index: number): Scannable {
+    return this.children[index];
+  }
+
+  addChild(child: Scannable) {
+    this.children.push(child);
+    this.sortChildren();
+  }
+
+  sortChildren() {
+    this.children.sort(this._sortScannables);
+  }
+
+  _sortScannables(a: Scannable, b: Scannable): number {
+    return a.priority - b.priority;
+  }
+}
+
+export class ScanningModel {
+  currentHighlight: Scannable;
+  currentSelection: Scannable;
+}
+
+export interface ScannableCollectionProvider extends Subscriber<ScanningModel> {
+  getScannableCollections(): ScannableCollection[];
+}
+
+@Injectable({
+  providedIn: 'root'
+})
+export class ScanningService {
+
+  private topLevelScannables: ScannableCollection;
+  private currentCollection: ScannableCollection;
+  private scanningModel: ScanningModel;
+  private currentSelectedIndex = 0;
+  private observers: ScannableCollectionProvider[];
+  private intervalId;
+
+  constructor(private configService: ConfigService) {
+    this.topLevelScannables = new ScannableCollection(0, '');
+    this.currentCollection = this.topLevelScannables;
+    this.scanningModel = new ScanningModel();
+    this.observers = [];
+
+    if (this.configService.scanningConfig.enabled) {
+      document.onkeydown = this.handleInteraction.bind(this);
+    }
+  }
+
+  getScanningModel(): Observable<ScanningModel> {
+    return new Observable<ScanningModel>(this.registerScannable);
+  }
+
+  registerScannable = (observer: ScannableCollectionProvider): Subscription => {
+    this.observers.push(observer);
+
+    this.topLevelScannables.getChildren().push(...observer.getScannableCollections());
+    this.topLevelScannables.sortChildren();
+    this.startScanning();
+
+    return new Subscription(() => this._unsubscribe(observer));
+  }
+
+  _unsubscribe(observer: ScannableCollectionProvider) {
+    if (this.observers.includes(observer)) {
+      this.observers.splice(this.observers.indexOf(observer), 1);
+    }
+    observer.getScannableCollections().forEach(scannable => {
+      const collection = <ScannableCollection> scannable;
+      const index = this.topLevelScannables.getChildren().indexOf(collection);
```

Comment:
It's probably not a good idea to access getChildren() directly. A removeChild method would possibly be better